--- gcc-6.2.0/gcc/config/s390/linux.h.orig	2016-11-11 23:15:14.289495938 -0500
+++ gcc-6.2.0/gcc/config/s390/linux.h	2016-11-11 23:20:25.854648405 -0500
@@ -63,6 +63,9 @@
 #define GLIBC_DYNAMIC_LINKER32 "/lib/ld.so.1"
 #define GLIBC_DYNAMIC_LINKER64 "/lib/ld64.so.1"
 
+#define MUSL_DYNAMIC_LINKER32 "/lib/ld-musl-s390.so.1"
+#define MUSL_DYNAMIC_LINKER64 "/lib/ld-musl-s390x.so.1"
+
 #undef  LINK_SPEC
 #define LINK_SPEC \
   "%{m31:-m elf_s390}%{m64:-m elf64_s390} \
