--- gcc-8.5.0/gcc/config/rs6000/sysv4.h.orig	2021-07-05 12:56:18.398813835 -0400
+++ gcc-8.5.0/gcc/config/rs6000/sysv4.h	2021-06-23 21:49:57.592408186 -0400
@@ -793,7 +793,7 @@
 #define GLIBC_DYNAMIC_LINKER "/lib/ld.so.1"
 #define UCLIBC_DYNAMIC_LINKER "/lib/ld-uClibc.so.0"
 #define MUSL_DYNAMIC_LINKER \
-  "/lib/ld-musl-powerpc" MUSL_DYNAMIC_LINKER_E "%{msoft-float:-sf}.so.1"
+  "/lib/ld-musl-powerpc" MUSL_DYNAMIC_LINKER_E "%{msoft-float|mcpu=854*|mspe:-sf}.so.1"
 #if DEFAULT_LIBC == LIBC_UCLIBC
 #define CHOOSE_DYNAMIC_LINKER(G, U, M) \
   "%{mglibc:" G ";:%{mmusl:" M ";:" U "}}"
--- gcc-8.5.0/gcc/config/rs6000/linux64.h.orig	2021-07-05 12:54:24.704854044 -0400
+++ gcc-8.5.0/gcc/config/rs6000/linux64.h	2021-07-05 12:56:50.010806923 -0400
@@ -426,7 +426,7 @@
 #endif
 
 #define MUSL_DYNAMIC_LINKER32 \
-  "/lib/ld-musl-powerpc" MUSL_DYNAMIC_LINKER_E "%{msoft-float:-sf}.so.1"
+  "/lib/ld-musl-powerpc" MUSL_DYNAMIC_LINKER_E "%{msoft-float|mcpu=854*|mspe:-sf}.so.1"
 #define MUSL_DYNAMIC_LINKER64 \
   "/lib/ld-musl-powerpc64" MUSL_DYNAMIC_LINKER_E "%{msoft-float:-sf}.so.1"
 
