diff -ur a/libgcc/unwind-pe.h b/libgcc/unwind-pe.h
--- a/libgcc/unwind-pe.h	2020-05-07 05:50:02.000000000 -0500
+++ b/libgcc/unwind-pe.h	2020-05-07 22:03:49.602635844 -0500
@@ -29,6 +29,15 @@
 #ifndef GCC_UNWIND_PE_H
 #define GCC_UNWIND_PE_H
 
+#ifdef __SH2__
+_Unwind_Ptr __attribute__((weak)) __gnu_Unwind_Find_got (_Unwind_Ptr);
+
+static inline _Unwind_Ptr _Unwind_gnu_Find_got (_Unwind_Ptr ptr)
+{
+    return __gnu_Unwind_Find_got (ptr);
+}
+#endif
+
 /* If using C++, references to abort have to be qualified with std::.  */
 #if __cplusplus
 #define __gxx_abort std::abort
